src/nyx/gpu/geomoments_central.cu

```diff
@@ -0,0 +1,234 @@
+#include <cub/cub.cuh>
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
+#include <builtin_types.h>
+#include "gpu.h"
+#include "../features/pixel.h"
+
+namespace Nyxus
+{
+    extern double* devPrereduce;      // reduction helper [roi_cloud_len]
+    extern double* devBlockSubsums;   // [whole chunks]
+    extern double* hoBlockSubsums;    // [whole chunks]
+};
+
+__device__ double pow_pos_int_central (double a, int b)
+{
+    if (b == 0)
+        return 1.0;
+    double retval = 1.0;
+    for (int i = 0; i < b; i++)
+        retval *= a;
+    return retval;
+}
+
+__global__ void kerCentralMoment (
+    double* d_prereduce,
+    const Pixel2* d_roicloud,
+    size_t cloudlen,
+    StatsInt base_x,
+    StatsInt base_y,
+    double origin_x, 
+    double origin_y,
+    int p,
+    int q)
+{
+    int tid = threadIdx.x + blockIdx.x * blockSize;
+    
+    if (tid >= cloudlen)
+        return;
+
+    d_prereduce[tid] = double(d_roicloud[tid].inten) 
+        * pow_pos_int_central (double(d_roicloud[tid].x - base_x) - origin_x, p) 
+        * pow_pos_int_central (double(d_roicloud[tid].y - base_y) - origin_y, q);
+}
+
+__global__ void kerCentralMomentWeighted (
+    double* d_prereduce,
+    const RealPixIntens* d_realintens,
+    const Pixel2* d_roicloud,
+    size_t cloudlen,
+    StatsInt base_x,
+    StatsInt base_y,
+    double origin_x,
+    double origin_y,
+    int p,
+    int q)
+{
+    int tid = threadIdx.x + blockIdx.x * blockSize;
+
+    if (tid >= cloudlen)
+        return;
+
+    d_prereduce[tid] = d_realintens[tid]
+        * pow_pos_int_central(double(d_roicloud[tid].x - base_x) - origin_x, p)
+        * pow_pos_int_central(double(d_roicloud[tid].y - base_y) - origin_y, q);
+}
+
+bool drvCentralMoment(
+    double& retval,
+    int p, int q,
+    const Pixel2* d_roicloud,
+    size_t cloudlen,
+    StatsInt base_x,
+    StatsInt base_y,
+    double origin_x,
+    double origin_y)
+{
+    int nblo = whole_chunks2(cloudlen, blockSize);
+    kerCentralMoment <<< nblo, blockSize >>> (Nyxus::devPrereduce, d_roicloud, cloudlen, base_x, base_y, origin_x, origin_y, p, q);
+
+    CHECKERR(cudaPeekAtLastError());
+    CHECKERR(cudaDeviceSynchronize());
+    CHECKERR(cudaGetLastError());
+
+    //=== device-reduce:
+    // Determine temporary device storage requirements
+    double* d_out = nullptr;
+    CHECKERR(cudaMalloc(&d_out, sizeof(double)));
+    void* d_temp_storage = NULL;
+    size_t   temp_storage_bytes = 0;
+    cub::DeviceReduce::Sum(d_temp_storage, temp_storage_bytes, Nyxus::devPrereduce/*d_in*/, d_out, cloudlen/*num_items*/);
+    // Allocate temporary storage
+    cudaMalloc(&d_temp_storage, temp_storage_bytes);

```

Comment:
Do we need to check the error status of cudaMalloc (and other allocations missing error check)? i.e. 

ok = cudaMalloc(&d_temp_storage, temp_storage_bytes);
CHECKERR(ok);

Reply:
We have 2 error-checking options here: (1) not bother of checking cudaMalloc() relying on internal error handling in Sum() since those functions are coupled (2) do error checking defensively forgetting that cudaMalloc() allocates a tiny chunk of memory and Sum() is the only point of failure. Opted for (2).